share/availability/discovery/discovery.go

```diff
@@ -96,26 +109,42 @@ func (d *Discovery) WithOnPeersUpdate(f OnUpdatedPeers) {
 
 // handlePeersFound receives peers and tries to establish a connection with them.
 // Peer will be added to PeerCache if connection succeeds.
-func (d *Discovery) handlePeerFound(ctx context.Context, topic string, peer peer.AddrInfo) {
-	if peer.ID == d.host.ID() || len(peer.Addrs) == 0 || d.set.Contains(peer.ID) {
+func (d *Discovery) handlePeerFound(ctx context.Context, peer peer.AddrInfo, cancelFind context.CancelFunc) {
```

Comment:
Ok, so I verified and `peer.AddrInfo` here, and it should always contain empty addresses and only the ID filled is set. 
All the references to [this](https://github.com/libp2p/go-libp2p-kad-dht/blob/master/providers/providers_manager.go#L242) do only put the ID, but not the actual addresses. Therefore, the RoutedHost must find the address behind the peers here, and we should have a higher Connect timeout.